sky/spot/spot_utils.py

```diff
@@ -232,56 +237,65 @@ def stream_logs_by_id(job_id: int, follow: bool = True) -> str:
+    with status_display:
+        while spot_status is None or not spot_status.is_terminal():
+            handle = global_user_state.get_handle_from_cluster_name(
+                cluster_name)
+            # Check the handle: The cluster can be preempted and removed from
+            # the table before the spot state is updated by the controller. In
+            # this case, we should skip the logging, and wait for the next
+            # round of status check.
+            if handle is None or spot_status != spot_state.SpotStatus.RUNNING:
+                status_help_str = ''
+                if (spot_status is not None and
+                        spot_status != spot_state.SpotStatus.RUNNING):
+                    status_help_str = f', status: {spot_status.value}'
+                logger.debug(
+                    f'INFO: The log is not ready yet{status_help_str}. '
+                    f'Waiting for {JOB_STATUS_CHECK_GAP_SECONDS} seconds.')
+                status_display.update('[bold cyan]Waiting for the spot job'
+                                      f'{status_help_str}.[/] It may take '
+                                      'a few minutes.')
+                time.sleep(JOB_STATUS_CHECK_GAP_SECONDS)
+                spot_status = spot_state.get_status(job_id)
+                continue
+            status_display.stop()
+            returncode = backend.tail_logs(handle,
+                                           job_id=None,
+                                           spot_job_id=job_id,
+                                           follow=follow)
+            if returncode == 0:
+                # If the log tailing exit successfully (the real job can be
+                # SUCCEEDED or FAILED), we can safely break the loop. We use the
+                # status in job queue to show the information, as the spot_state
+                # is not updated yet.
+                job_statuses = backend.get_job_status(handle, stream_logs=False)
+                job_status = list(job_statuses.values())[0]
+                assert job_status is not None, 'No job found.'
+                if job_status != job_lib.JobStatus.CANCELLED:
+                    logger.info(f'Logs finished for job {job_id} '
+                                f'(status: {job_status.value}).')
+                    break
+                # The job can be cancelled by the user or the controller (when
+                # the cluster is partially preempted).
+                logger.info('INFO: (Log streaming) Job is cancelled. Waiting '
+                            'for the status update in '
+                            f'{JOB_STATUS_CHECK_GAP_SECONDS} seconds.')
+            else:
+                logger.debug(
+                    f'INFO: (Log streaming) Got return code {returncode}. '
+                    f'Retrying in {JOB_STATUS_CHECK_GAP_SECONDS} seconds.')
+            status_display.update('[bold cyan]Waiting for the spot job[/]')
```

Comment:
Q: I can't seem to trigger this log with sky spot launch sleep 10000.

Add "to start"?

Reply:
This only happens when the cluster is preempted, i.e. the `tail_logs` returns nonzero.